### DISCUSSION TITLE
Problems with connection and verification in Access c# databases

### DISCUSSION THREAD

**Discussion by Nlech3** [linq2db/linq2db#4538] [Q&A] [answered]
For a student project I need to create an application with an Access database, but I don't know anything about databases. Can anyone explain what the problem is?

private void button1_Click(object sender, EventArgs e)
{

    DataPerson.Full_name = textBox1.Text;
    DataPerson.Address = textBox2.Text;
    DataPerson.Contract_number = textBox3.Text;
    DataPerson.Telephone = textBox4.Text;

    string connectionString = "Provider=Microsoft.ACE.OLEDB.12.0;Data Source=D:\\IT\\WindowsFormsApp1\\WindowsFormsApp1\\Microsoft Access Database.accdb";

    using (OleDbConnection connection = new OleDbConnection(connectionString))
    {
        try
        {
            connection.Open();

            string query = "SELECT 1-3 FROM Table1 ФИО = @ФИО AND Адрес = @Адрес AND Номер_Договора = @Номер_Договора AND Телефон = @Телефон";

            OleDbCommand command = new OleDbCommand(query, connection);
            command.Parameters.AddWithValue("@ФИО", DataPerson.Full_name);
            command.Parameters.AddWithValue("@Адрес", DataPerson.Address);
            command.Parameters.AddWithValue("@Номер_Договора", DataPerson.Contract_number);
            command.Parameters.AddWithValue("@Телефон", DataPerson.Telephone);
            int count = Convert.ToInt32(command.ExecuteScalar());

            if (count > 3)
            {
                MessageBox.Show("Успешная авторизация!");
            }
            else 
            {
                MessageBox.Show("Неудачная авторизация или отсутствие аккаунта!");
            }
        }
        catch (Exception ex)
        {
            MessageBox.Show("Error: " + ex.ToString());
        }
        finally
        {
            connection.Close();
        }
    }
}

![Снимок экрана 2024-06-18 020823](https://github.com/linq2db/linq2db/assets/173110123/49da0e9c-3a1d-4ab0-956f-6e499ae5c1bd)

**Comment by MaceWindu** [comment, marked as answer]
You have syntax error in your query.

If you format your query it should be obvious that you miss `WHERE` keyword before query filter:

```sql
SELECT 1-3
    FROM Table1
    WHERE -- missing part
    ФИО = @ФИО AND Адрес = @Адрес AND Номер_Договора = @Номер_Договора AND Телефон = @Телефон
```

Check Access reference here https://learn.microsoft.com/en-us/office/client-developer/access/desktop-database-reference/select-statement-microsoft-access-sql

**Reply by MaceWindu** [reply]
BTW, 
1. your select expression `1-3` is strange. It's same as `-2` and will never be `> 3`
2. query will return 0 records if filter returns false. Better to use `COUNT` aggregate instead: `SELECT COUNT(*)` and check resulting number being equal to 1 (record)
